File: python/156. Binary Tree Upside Down.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def upsideDownBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if not root or not root.left:
            return root
        l = self.upsideDownBinaryTree(root.left)
        r = l
        while r.right:
            r = r.right
        root, r.left, r.right = l, root.right, TreeNode(root.val)
        return root

File: cpp/151-200/156. Binary Tree Upside Down.cc
class Solution {
public:
    TreeNode* upsideDownBinaryTree(TreeNode* root) {
        return helper(root, NULL);
    }
    TreeNode* helper(TreeNode *root, TreeNode *parent) {
        if (!root)
            return NULL;
        if (!root->left && !root->right) {
            root->left = parent ? parent->right : NULL;
            root->right = parent;
            return root;
        }
        TreeNode *left = helper(root->left, root);
        root->left = parent ? parent->right : NULL;
        root->right = parent;
        return left;
    }
};